### PR TITLE
Allow up to 5 sessions for users

### PR DESCRIPTION
**Pull request by DevDonut**
Allow up to 5 sessions for users, so you won't be logged out of the admin site when logging in to the regular site and vice versa.

### CODE REVIEW

**Comment by skamoen** [comment]
An Optional of a List is is bit much. If you omit the Optional, a query with no results will just give you an empty list, making the whole thing a lot easier.

**Comment by dsluijk** [comment]
This feels like an hack, it might be better to replace it with something like JWT tokens? Should be lighter on the database too.

**Comment by DevDonut** [comment]
> This feels like an hack, it might be better to replace it with something like JWT tokens? Should be lighter on the database too.

@dsluijk i feel like this would require a complete overhaul of the Authentication service as currently implemented. I myself have no experience with JWT tokens, so I'm not sure how much work that would be.

File: src/main/java/ch/wisv/areafiftylan/security/authentication/AuthenticationServiceImpl.java [hunk 1]
```diff
@@ -58,8 +61,11 @@ public AuthenticationServiceImpl(AuthenticationTokenRepository authenticationTok
     public String createNewAuthToken(String email) {
         User user = userService.getUserByEmail(email);
 
-        // Delete the old Token
-        authenticationTokenRepository.findByUserEmail(email).ifPresent(authenticationTokenRepository::delete);
+        // Delete the old Tokens if there are more than 4.
+        List<AuthenticationToken> existingTokens = authenticationTokenRepository.findByUserEmail(email);
+        if (existingTokens.size() >= 4) {
+            authenticationTokenRepository.deleteAll(existingTokens);
```
Comment (martijnjanssen) [inline src/main/java/ch/wisv/areafiftylan/security/authentication/AuthenticationServiceImpl.java:67]: So if there are more than 4 tokens, we're deleting them all? Makes more sense to only delete the oldest right? Deleting them all introduces a sort of random factor for the users. One time they are still logged in, and the other time all devices are signed out.

**Review by martijnjanssen — COMMENTED** [review]
Cool feature! This is quite the improvement

**Comment by martijnjanssen** [comment]
> This feels like an hack, it might be better to replace it with something like JWT tokens? Should be lighter on the database too.

This is definitely not a hack, the solution proposed is reasonable. Switching to JWT also has many downsides (e.g. elaborate checking logic for users that still might have valid token while the user is invalid/banned), apart from the fact that re-implementing the authentication makes little sense.

File: src/main/java/ch/wisv/areafiftylan/security/authentication/AuthenticationServiceImpl.java (hunk 1 above)
Comment (DevDonut) [reply src/main/java/ch/wisv/areafiftylan/security/authentication/AuthenticationServiceImpl.java:67]: That's a solid point! I will change it.

**Review by skamoen — APPROVED** [review]

**Pull request merged**